src/selectors/linked-nodes.js

```diff
@@ -17,36 +15,79 @@ export const getCentralNode = createSelector(
 );
 
 /**
- * Recursively search through the edges data for ancestor and descendant nodes
+ * Gets a map of nodeIDs to the visible source edges and target edges
  * @param {Array} edges
+ */
+export const getVisibleEdgesByNode = createSelector(
+  [getVisibleEdges],
+  edges => {
+    const sourceEdges = {};
+    const targetEdges = {};
+
+    for (const edge of edges) {
+      sourceEdges[edge.target] = sourceEdges[edge.target] || [];
+      sourceEdges[edge.target].push(edge);
+    }
+
+    for (const edge of edges) {
+      targetEdges[edge.source] = targetEdges[edge.source] || [];
+      targetEdges[edge.source].push(edge);
+    }
+
+    return { sourceEdges, targetEdges };
+  }
+);
+
+/**
+ * Finds all visible successors for the given nodeID in given direction
```

Comment:
So is the problem with the old one, that it will revisit ascendants/descendants of nodes that are linked to twice, e.g. in situations like this?
```
2     2
 \   /
   2
 /   \
1     1
 \   /
   A
```
(where A is the original hovered node, and the numbers are nodes with the number referring to how many times each one gets visited)

Reply:
If I understand it correctly, the original version appears to loop over every edge for every edge visited, so it degrades to `O(edges^2)` in the worst case (e.g. highly connected graphs). It can also end up traversing the same edges multiple times (since paths may intersect). Both of these problems should now be avoided with this approach, hopefully that makes sense?

Reply:
~I think only the second case is true~

omg you're right, it does! For some reason I thought it only traversed to linked nodes but you're right, I'd forgotten about the loop. Terrible. I'd been meaning to optimise this but it hadn't occurred to me quite how bad this part was.